### ISSUE TITLE
Virtualised menu items 

### ISSUE THREAD

**Issue by Confidenceman02** [Confidenceman02/elm-select#163]
Elm-select will render all the items you give it and if it's thousands then things can lag. There is nothing unusual about this, its more so that the DOM can get slow with huge amounts of elements.

For the most part, folks wont really run in to this issue, especially when they are careful about the amount of items they are loading, but none the less I think it's worth thinking about performance and how elm-select might provide options for consumers when they do indeed want to render huge amounts of items.

## Solutions
- Consumers to be careful about the amount of data they are loading in to elm-select. Elm-select takes no responsibility or ownership. (lowest effort)
- Provide a list virtualisation solution internally. This would mean elm-select handles the virtualisation entirely. (high effort).
- Provide an example for consumers to handle virtualisation with some elm-select help (high effort).

## Resources 
- [VIDEO: react-virtualized](https://www.youtube.com/watch?v=UrgfPjX97Yg)
- [SOURCE: react-virtualized](https://github.com/bvaughn/react-virtualized/tree/master)
- [react-window](https://github.com/bvaughn/react-window)
- [translate vs top for positioning](https://stackoverflow.com/questions/20180735/what-is-the-difference-between-translate-and-top-in-css)
- [CodeSandBox example react-window](https://codesandbox.io/s/react-window-list-padding-dg0pq)

**Comment by hiddenroadss** [comment]
Hi,
Have you tried to implement it yet? I realised we need this feature in our app (have pretty long lists at some pages). So now I'm trying to understand how to fix it and virtual scroll came to my mind.
If you are busy with more important tasks now, maybe you would give some advice how to add it to existing select code?(like fork it)

**Comment by Confidenceman02** [comment]
Hi there!

> Have you tried to implement it yet?

I haven't yet as it was a problem that I didn't encounter and was waiting for the right time, now might be that time! :)

> I realised we need this feature in our app (have pretty long lists at some pages). So now I'm trying to understand how to fix it and virtual scroll came to my mind.

Virtual lists definitely are a handy thing to have if you can't avoid incremental rendering of list items. Just to better understand your requirements and perhaps offer an alternative solution before going down the virtual scroll rabbit hole.

**Do you have an indication of how long the lists are generally?**

**Would your UI be able to handle asynchronous list items?**

For example, elm-select is really optimized to find list items via search rather than scroll especially when the lists are long. 

Is it possible for your UI load menu items asynchronously on search? Elm-select has search triggers and an async mode to visually show that items are being loaded. If that is feasible I can provide an example of how you could hook that up if you need it.



> If you are busy with more important tasks now, maybe you would give some advice how to add it to existing select code?(like fork it)

If you feel you 100% need a virtual list I am more than happy to start a decent investigation and work with you to get a good solution.

What do you think?

**Comment by hiddenroadss** [comment]
Thanks for your reply !(github didn't notify me for some reason)

So answering your questions:

> Do you have an indication of how long the lists are generally?
Yes, it depends on a date range, if user loads data for 1 day - it might be 10-50 items, but if we load data for a week - it might be up to 800. which is influence performance and UX. (I'm not even talking about trying to upload month of data:D)

> Would your UI be able to handle asynchronous list items?
If I understood you right, you mean loading new items from server while user scroll/search, correct? It is possible, definitely, but for now I'm not sure I can convince backend guys to implement this feature, because they have another job to do.
That is why I was thinking about front-end solution to this problem (of course, it won't be as good as with backend, but we need to start somewhere). And the most obvious enhancment I've thought of was virtual lists, just because DOM nodes takes a lot of memory
What do you think about it?

**Comment by Confidenceman02** [comment]
> but if we load data for a week - it might be up to 800. which is influence performance and UX.

Got it! Yea 800 items would mean a slow DOM for sure.

> If I understood you right, you mean loading new items from server while user scroll/search

Not while scroll, but when search. It would mean the same/similar backend work though so either way if your team doesn't have capacity to build it then you are equally stuck.

> And the most obvious enhancement I've thought of was virtual lists, just because DOM nodes takes a lot of memory

Yea, there is a lot going on and making it feel snappy definitely needs virtual scroll when your rendering that many nodes, if you can't lazy load them via an API.

I started this work a few days ago in a `virtual-scroll` branch. I'm just building the windowing algorithm now so I'll let you know how it goes. I'm confident we will get a good solution :)

You will hear back from me by end of the week.

**Comment by Confidenceman02** [comment]
Howdy there @hiddenroadss !

I have it working!

![virtual-scroll](https://github.com/user-attachments/assets/2d18bd16-335d-42df-87b3-4c3169cf0c2f)


The code is an absolute mess but I'm confident the approach works. 

I'm going to clean it up and perhaps do some performance testing but looking very promising! 

Hang in there!

**Comment by hiddenroadss** [comment]
Wow, that's fantastic! I'm looking forward to checking your code after a clean up. I'm curious how to implement such thing using elm

btw, how do you test performance of elm app? using e2e tests? (saw some in your test folder) 
or do you use manual tests also?)

**Comment by Confidenceman02** [comment]
> Wow, that's fantastic! I'm looking forward to checking your code after a clean up. I'm curious how to implement such thing using elm

My implementation is by no means generic, it's really tailored to elm-select which can be a bit of a beast to work with.

> btw, how do you test performance of elm app? using e2e tests? (saw some in your test folder)
> or do you use manual tests also?)

The e2e tests are mostly there to ensure nothing regresses and all the complex cases are covered.

I use dev tools a lot for performance testing and see how the browser is coping with the scripting and the rendering.

For virtual scroll the painting and rendering will look pretty good even with huge lists, but the scripting might take a hit. It's a tradeoff! 

There are still performance gains on the table, but can iterate on it over time. For ultimate performance if it's required we can always use elm-select's `jsOptimize` flag. Means we can perform some heavy lifting outside of Elm's update cycle but its a "Only if absolutely necessary" kind of thing. Better to keep it all Elm.

I've found some gnarly bugs as a result of virtual scroll so going to get through them then check back in with you.

**Comment by Confidenceman02** [comment]
Ready for release. 

Feel free to have a look at the PR. It's a bit  of a beast but the magic is happening in `viewVirtual` in `Select.elm`.

Get get a good sense of what is happening you can compare it to the normal `view` function. 

Happy to explain any code that you want more info on also!

**Comment by Confidenceman02** [comment]
released in [11.0.0](https://package.elm-lang.org/packages/Confidenceman02/elm-select/11.0.0/)

Let me know how you go!

**Comment by hiddenroadss** [comment]
Great news, I will finish my current task, will try it out,  and will share the progress with you)

**Comment by hiddenroadss** [comment]
@Confidenceman02 , hello)

> And only the Single variant is supported, not Multi or any of the others. For now!

All our heavy selects are multi-selects. They work like a filters to save space, we have filter side-bar and a bunch of selects there)
So I'm curious,  whether or not it will be burdensome  for you to add it to Multi variant?

**Comment by Confidenceman02** [comment]
> @Confidenceman02 , hello)
> 
> > And only the Single variant is supported, not Multi or any of the others. For now!
> 
> All our heavy selects are multi-selects. They work like a filters to save space, we have filter side-bar and a bunch of selects there) So I'm curious, whether or not it will be burdensome for you to add it to Multi variant?

Hi there!

Ah I see!

I think `Multi` should just work so I will test it out this weekend and let you know how I go.

**Comment by hiddenroadss** [comment]
@Confidenceman02 , hello again)
It is not related to this issue (but I don't know where I can ask you something like this), I'm curious about your experience in handling big forms with elm and elm-select. I've seen example where you use record for each select (like {items, selectState, selectedItems}) and it fixes problem with messy `Model`, but on the other hand updating nested records is pretty hard in elm, so it add some boilerplate. It is ok to have 1 level of nesting, but if you want something like
```elm
type alias Model = {
addState : {inpSelect: {items, selectState, selectedItems}}
editState: {...inputs}
}
```
then it is already 2 levels of nesting and updating it seems really hard .

Long story short, what are your thoughts on this matter?

**Comment by Confidenceman02** [comment]
Howdy! 

I think this is a better question for the Elm [slack](https://elm-lang.org/community/slack) channel, or [Elm discourse forum](https://discourse.elm-lang.org/) or the Elm discord server.

This is more of a Elm pattern question than elm-select question.

But, if you create a working example of exactly what you mean, i'll happily help you out.

**Comment by Confidenceman02** [comment]
Just did some testing and yep looks like Multi just works.

I'm going to add tests for it then will do a release.

**Comment by Confidenceman02** [comment]
released in [11.1.0](https://package.elm-lang.org/packages/Confidenceman02/elm-select/11.1.0/)

**Comment by Confidenceman02** [comment]
let me know how you go!

**Comment by hiddenroadss** [comment]
I've changed our selects, it was really easy. Now I'm waiting till we deploy it in prod and check if everything is fine (I don't have such huge mock data in dev env). 
So I will come back with updates a bit later

**Comment by hiddenroadss** [comment]
I've just scrolled through 12k items select and it was smooth and lag-free experience :)

**Comment by Confidenceman02** [comment]
> I've just scrolled through 12k items select and it was smooth and lag-free experience :)

WOW!

I only tested witth 2K. That is epic!

You happy for this issue to be closed?

**Comment by hiddenroadss** [comment]
Yes, thank you for your help!
